### ISSUE TITLE
feat: allow declarations/entryComponents on createService

### ISSUE THREAD

**Issue by IlCallo** [ngneat/spectator#106]
Hi, I'm trying to test a service using MatDialog to open a component.
I want to mock the component, because I'm only interest in the service stuff, not really what the component does.
Problem is I don't have `declarations` and `entryComponents` on `createService` method, so I cannot specify a mock for the component.

I tried adding a dummy module and mocking it with ng-mocks like this

```ts
@NgModule({
  declarations: [MyComponent],
  entryComponents: [MyComponent]
})
class DummyModule {}

const spectator = createService({
  service: MyService,
  imports: [NoopAnimationsModule, MatDialogModule, MockModule(DummyModule)]
});
```

It doesn't work, keep asking for `MyComponent` to be defined as an entry component.
Is it possible to allow the method to accept also `declarations` and `entryComponents`?

**Comment by NetanelBasal** [comment]
Can you share the service, please?

**Comment by IlCallo** [comment]
```ts
@Injectable({
  providedIn: 'root'
})
export class MyService {
  constructor(private dialog: MatDialog) {}

  open(inputItem$: Observable<Item>, handlers: Partial<OutputHandlers> = {}): MatDialogRef<MyComponent> {
    const dialogRef = this.dialog.open(MyComponent);

    inputItem$.subscribe(item => (dialogRef.componentInstance.item = item));

    each(handlers, (handler, key) => (dialogRef.componentInstance as any)[key].subscribe(handler));

    return dialogRef;
  }
}
```

The service is basically a proxy to the component, to use it both via Dialog and in templates.

**Comment by NetanelBasal** [comment]
Yes, I see. The issue is this `this.dialog.open(MyComponent);` line.

**Comment by NetanelBasal** [comment]
We should allow passing `entryComponents` and `declarations` here: https://github.com/NetanelBasal/spectator/blob/master/projects/spectator/src/lib/service.ts#L31

**Comment by IlCallo** [comment]
Yes, probably it could be a good idea to keep a single basic interface for both createComponent and createService and then specialize it for extra options of the two methods.

Probably the base should be derived from https://github.com/NetanelBasal/spectator/blob/9edad7d1c68cf399768b5d7cbe4e7f21f598cfca/projects/spectator/src/lib/config.ts#L20-L30

I also opened an issue on ng.mocks for (probably) faulty module mocking: https://github.com/ike18t/ng-mocks/issues/51

**Comment by NetanelBasal** [comment]
Try 3.8.0, please.

**Comment by IlCallo** [comment]
Tryed, apparently it didn't help much.
I tryed it both using ng-mocks and with the real component and its dependencies.

```ts
  const spectator = createService({
    service: MyService,
    imports: [
      NoopAnimationsModule,
      CommonModule,
      MatDialogModule,
      // ... dependencies of MyComponent
    ],
    providers: [
      // ... dependencies of MyComponent
    ],
    declarations: [MyComponent],
    entryComponents: [MyComponent]
  });
```

```ts
  const spectator = createService({
    service: MyService,
    imports: [NoopAnimationsModule, MatDialogModule],
    declarations: [MockComponent(MyComponent)],
    entryComponents: [MockComponent(MyComponent)]
  });
```

Still same error
```
No component factory found for MyComponent. Did you add it to @NgModule.entryComponents?

      at noComponentFactoryError (../../../packages/core/src/linker/component_factory_resolver.ts:17:17)
      at CodegenComponentFactoryResolver.resolveComponentFactory (../../../packages/core/src/linker/component_factory_resolver.ts:62:13)
      at CdkPortalOutlet.Object.<anonymous>.CdkPortalOutlet.attachComponentPortal (../../node_modules/@angular/src/cdk/portal/portal-directives.ts:130:39)
      at MatDialogContainer.Object.<anonymous>.MatDialogContainer.attachComponentPortal (../../node_modules/@angular/src/lib/dialog/dialog-container.ts:116:31)
      at MatDialog.Object.<anonymous>.MatDialog._attachDialogContent (../../node_modules/@angular/src/lib/dialog/dialog.ts:267:42)
      at MatDialog.Object.<anonymous>.MatDialog.open (../../node_modules/@angular/src/lib/dialog/dialog.ts:138:28)
      at MyService.Object.<anonymous>.MyService.open (src/lib/my.service.ts:24:35)
      at src/lib/my.service.spec.ts:48:13
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (../../node_modules/zone.js/dist/zone.js:391:26)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (../../node_modules/zone.js/dist/proxy.js:129:39)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (../../node_modules/zone.js/dist/zone.js:390:52)
      at Zone.Object.<anonymous>.Zone.run (../../node_modules/zone.js/dist/zone.js:150:43)
      at Object.testBody.length (../../node_modules/jest-preset-angular/zone-patch/index.js:52:27)
```

Any idea on how can I debug this further?

**Comment by NetanelBasal** [comment]
Can you create a simple demo of this, please? stackblitz or new repo, so I can investigate further.

**Comment by benelliott** [comment]
@IlCallo I expect the problem is that in your service you are directly referencing the class of the unmocked component. When you ask `MatDialog` to instantiate this it will complain because this component cannot be found in the injector tree (you have mocked it, but your service will not automagically point to the mocked component instead).

If I'm correct, this is more a limitation of your codebase than your unit tests. One option could be for you to stub `MatDialog` so that it doesn't actually try to resolve the component.

**Comment by IlCallo** [comment]
If that's the case, shoudn't the third way I tried work?

This one

```ts
const spectator = createService({
    service: MyService,
    imports: [NoopAnimationsModule, MatDialogModule],
    providers: [{ provide: MyComponent, useClass: MockComponent(MyComponent) }],
    entryComponents: [MyComponent]
  });
```

Given that I'm providing the mocked class using the original class token.

Next week, hopefully, I'll try to refactor the codebase to stub `MatDialog` or creare a repro (if the refactor won't work)

**Comment by benelliott** [comment]
@IlCallo I think that would only work if the service that instantiates the dialog pulls the component class from the injector rather than by directly pointing to the class file.

**Comment by IlCallo** [comment]
@NetanelBasal is there a Spectator-ready stackblitz from which to create a repro?

**Comment by NetanelBasal** [comment]
https://stackblitz.com/edit/angular-testing-spectator-ivrsbj

**Comment by IlCallo** [comment]
Here the repro.
https://stackblitz.com/edit/angular-testing-spectator-td7onb

I had to update some dependencies and I removed unusued code.

**Comment by NetanelBasal** [comment]
You can use the `ANALYZE_FOR_ENTRY_COMPONENTS` provider to register it. 

```ts
  const spectator = createService({
    service: HelloService,
    imports: [NoopAnimationsModule, MatDialogModule],
    providers: [{
      provide: ANALYZE_FOR_ENTRY_COMPONENTS,
      multi: true,
      useValue: HelloComponent
    }],
    declarations: [HelloComponent],
  });
```

**Comment by IlCallo** [comment]
Updated repro: it works if `ng-mocks` isn't used to mock the component, this work also just by providing components into `entryComponents` in some scenarios.
It could be a nice add to internally convert what is declared into `entryComponents` to that provider-notation.

Unfortunately, this doesn't totally resolve the use case.
In the repro, the component is super simple and without nested components, which isn't the case in the wast majority of cases (including mine), so mocking the component is a must-do (unless changing the schema, which I consider a bad practice).

**Comment by NetanelBasal** [comment]
Yes. I guess it's something with the ng-mocks implementation.

**Comment by benelliott** [comment]
From glancing at the ng-mocks source code (https://github.com/ike18t/ng-mocks/blob/master/lib/mock-component/mock-component.ts#L86) it looks like the mocked component only `provide`s itself in its own injector. This isn't sufficient for `MatDialog` because it will be injected wherever `MatDialogModule` has been imported, so won't see the component declaration.

**Comment by ike18t** [comment]
I don't have any experience using Angular Material but I do see that I can get the component's providers via the MockDirectiveResolver...that may be a slippery slope of needing to know what providers to mock vs what not to mock though.

**Comment by IlCallo** [comment]
Apparently being a ng-mocks issue, let's continue the discussion on https://github.com/ike18t/ng-mocks/issues/51 and let's see if it can and should be fixed

**Comment by dirkluijk** [comment]
Support for service factories (`createServiceFactory` / `SpectatorService`) with entry components has just been released in 4.0.0. Please consult the updated README.

Feel free to open a new ticket if you have ideas or questions.
